Added cosmos shell extension.

mkrueger opened the pull request:
## Add Cosmos DB Shell integration

Integrates the external [Cosmos DB Shell](https://aka.ms/cosmosdbshell) (`cosmosdbshell` CLI / .NET tool) into the extension. Users can launch a connected shell from the resource tree and optionally expose it as an MCP server for GitHub Copilot.

### Highlights

- **New command** `NoSQL: Launch Cosmos DB Shell` with context-menu entries on NoSQL database/container items (gated by `vscodeDatabases.cosmosDBShellSupportEnabled`).
- **Unified connect/launch flow** that reuses unconnected terminals opened from the command palette, with early-exit monitoring for clear error messages when the shell is missing.
- **Authentication**: connection string passthrough, emulator (no creds), Entra ID (token via `getAccessTokenForVSCode` with fallback handling), and managed identity.
- **Cross-platform command resolution** (`cosmosDBShellCommandResolver.ts`): walks `PATH`/`PATHEXT` on Windows and unwraps `dotnet tool` `.cmd`/`.bat` shims; passthrough on macOS/Linux. Honors a configured absolute path and surfaces actionable errors.
- **MCP server provider** (`cosmosDbShellMcpProvider`) that exposes the shell over HTTP on `127.0.0.1` at a configurable port and refreshes when shell settings change.
- **Language contribution** for `.csh` (`cosmosdbshell`) with basic bracket/comment/auto-closing config.

### New settings

| Setting | Default | Purpose |
| --- | --- | --- |
| `cosmosDB.shell.path` | `null` | Optional explicit path to the `cosmosdbshell` executable. |
| `cosmosDB.shell.MCP.enabled` | `false` | Expose the shell as an MCP server. |
| `cosmosDB.shell.MCP.port` | `6128` | Port for the MCP HTTP endpoint (127.0.0.1). |

### Notes

- Uses the proposed `mcpServerDefinitionProviders` API (typings vendored in `vscode.proposed.d.ts`).
- Adds `vscode-languageclient ~9.0.1`.
- Requires the user to install `cosmosdbshell` separately; degrades gracefully when not found.
- Unit tests cover Windows command resolution and MCP endpoint formatting.

sevoku commented on package.json:
```diff
@@ -1289,6 +1301,24 @@
+          },
+          "cosmosDB.shell.MCP.enabled": {
+            "type": "boolean",
+            "default": true,
```
```suggestion
            "default": false,
```
Let's turn this off by default

sevoku commented on src/cosmosShell/CosmosShellExtension.ts:
```diff
@@ -0,0 +1,123 @@
+    const isCosmosShellInstalled: boolean = isCosmosShellSupportEnabled();
+
+    if (!isCosmosShellInstalled) {
+        void vscode.window.showErrorMessage(
```
Let's add a "Settings" button which opens the settings URL: "vscode://settings/cosmosDB.shell.path"

mkrueger replied:
fixed.

sevoku commented on package.json:
```diff
@@ -916,6 +922,12 @@
+        {
+          "//": "[Database] Launch Cosmos Shell",
+          "command": "cosmosDB.launchCosmosShell",
+          "when": "view =~ /azure(ResourceGroups|Workspace|FocusView)/ && (viewItem =~ /treeitem[.]database(?![a-z.\\/])/i || viewItem =~ /treeitem[.]container(?![a-z.\\/])/i) && viewItem =~ /experience[.](table|cassandra|core)/i",
```
Let's add this to root account nodes as well, and we should remove `table` and anything else besides `core`

sevoku commented on src/cosmosShell/CosmosShellExtension.ts:
```diff
@@ -0,0 +1,123 @@
+
+    terminal.show();
+    if (node.model.container) {
+        terminal.sendText('cd ' + node.model.database.id + '/' + node.model.container.id, true);
```
Shouldn't this pass a CR at the end? For me the cd command doesn't get executed but only entered:
<img width="358" height="159" alt="image" src="https://github.com/user-attachments/assets/bea40cfd-4008-4e3a-a42f-c67e9961c896" />

mkrueger replied:
On windows it works without.  Interesting.

mkrueger committed:
Added cosmos shell extension.#2536

        - Implemented Cosmos Shell extension with MCP server integration
        - Added credential support for Cosmos Shell (including Entra ID)
        - Added MCP configuration settings
        - Added LSP server support
        - Added l10n localization strings
        - Added settings button to error messages
        - Fixed various configuration and import issues

mkrueger committed:
Fixed mcp option.

mkrueger committed:
Connect/Launch command merged.

Multiple shells are launched for several accounts. Re-used is just the shell for open accounts. So usually there is just 1 shell.

But multiple can be up - let's say for emulator.

mkrueger committed:
Fixed osx cosmos shell launch.

mkrueger committed:
Improved error message when cosmos shell path is configured but not found

mkrueger commented on src/cosmosShell/CosmosShellExtension.ts:
```diff
@@ -0,0 +1,441 @@
+        return;
+    }
+
+    const rawConnectionString = node.model.accountInfo.endpoint;
```
renamed.

mkrueger commented on src/cosmosShell/CosmosShellExtension.ts:
```diff
@@ -0,0 +1,441 @@
+        return false;
+    }
+}
+const McpServerName = 'cosmosdb-shell-mcp-server';
```
+1 sounds better.

mkrueger committed:
Some fixes from PR request.

sevoku committed:
fix: update regex to match both 'CosmosShell' and 'CosmosDBShell' in version check

sevoku committed:
fix: update terminal creation to use new API for shell arguments in launchCosmosShell

sevoku committed:
feat(shell): implement early exit monitoring for Cosmos Shell terminals and add more logging

sevoku committed:
fix(shell): skip passing credentials for emulator connections in launchCosmosShell

sevoku committed:
fix(shell): rename references from Cosmos Shell to Cosmos DB Shell in localization and code

sevoku committed:
feat(shell): add MCP server handling and notifications for Cosmos DB Shell

sevoku committed:
fix(shell): add MCP notification and error handling for missing Cosmos DB Shell

sevoku committed:
feat(shell): enhance Cosmos Shell launch with Entra ID credential support and fallback token handling

sevoku committed:
feat(shell): enhance Cosmos Shell credential handling for emulator and managed identities

mkrueger committed:
Fix Cosmos Shell command resolution on Windows

mkrueger committed:
Remove invalid 'when' clause from launchCosmosShell command declaration

mkrueger committed:
Remove duplicate updateTerminalContext function

mkrueger committed:
Refresh MCP server definition when shell settings change

mkrueger committed:
Cache isCosmosShellSupportEnabled result, invalidate on config change

mkrueger committed:
Reuse command-launched Cosmos DB Shell terminal on node-based launch

mkrueger committed:
Skip unconnected terminal reuse for Entra accounts to preserve fallback token

mkrueger committed:
Regenerate l10n bundle after rebase

mkrueger committed:
Remove stale eslint-disable for unknown rule

mkrueger committed:
Rename cosmosShell to cosmosDBShell; make command resolver platform-independent

- Rename folder src/cosmosShell -> src/cosmosDBShell, files and identifiers (CosmosShell* -> CosmosDBShell*, cosmosshell -> cosmosdbshell).
- Rename env vars COSMOS_SHELL_ACCOUNT_KEY/COSMOS_SHELL_TOKEN -> COSMOSDB_SHELL_ACCOUNT_KEY/COSMOSDB_SHELL_TOKEN.
- Update VS Code contributions (command id, context keys, language id, NLS keys).
- Use path.win32.* in the Windows branch of cosmosDBShellCommandResolver so unit tests pass on Linux CI.

mkrueger committed:
Merge branch 'main' into dev/mkrueger/cosmos_shell

mkrueger committed:
Use IPv4 loopback for Cosmos DB Shell MCP endpoint

mkrueger committed:
Merge branch 'main' into dev/mkrueger/cosmos_shell

mkrueger committed:
Merge branch 'main' into dev/mkrueger/cosmos_shell

mkrueger committed:
Merge branch 'main' into dev/mkrueger/cosmos_shell

sevoku reviewed (commented):
Nice, works much better now with the latest Shell build! Will approve after BugBash if no new issues come up.

mkrueger committed:
Improve Cosmos DB Shell install flow for new users

- Show command on tree items even when shell isn't installed
- Detect .NET SDK and offer guided install via dotnet tool
- Stream install output, detect success/failure, auto-relaunch
- Chain SDK install -> tool install via .NET Install Tool extension
- Offer Reload Window when PATH isn't refreshed yet

mkrueger committed:
Always reconnect when reusing Cosmos DB Shell terminal

Fixes AzureCosmosDB/cosmosdb-shell-preview#24. Track per-terminal launch state (endpoint + auth mode + tenant/MI client id) instead of just the endpoint, and always send 'connect <endpoint>' before navigating an existing terminal so a previously disconnected shell is re-attached. Reuse only when launch-time env credentials (account key / Entra fallback token) match; otherwise spawn a new shell so the right env vars get baked into the process.

mkrueger committed:
Use -k instead of --k for Cosmos DB Shell kickoff command

mkrueger committed:
Remove now-stable MCP API additions from vscode.proposed.d.ts

mkrueger committed:
Auto-continue Cosmos DB Shell install after .NET SDK acquisition

After acquiring the .NET SDK via the .NET Install Tool, PATH may not yet be
refreshed in the current VS Code session, so 'dotnet --list-sdks' on a bare
'dotnet' lookup fails and the install flow stalls silently. Plumb the
returned dotnetPath through isDotNetSdkInstalled and the dotnet tool install
spawn, and auto-chain into the Cosmos DB Shell install + launch instead of
asking the user to confirm again.

sevoku committed:
Enhance .NET SDK installation logic for Cosmos DB Shell with version checks and improved error handling

sevoku committed:
Merge origin/main into dev/mkrueger/cosmos_shell

Resolved conflicts:
- src/tree/cosmosdb/CosmosDBAccountAttachedResourceItem.ts: kept main's spread-based return shape; kept branch's getThemeAgnosticIconURI call
- src/tree/cosmosdb/CosmosDBAccountResourceItem.ts: same as above
- package-lock.json: regenerated from main

Migrated src/cosmosDBShell/cosmosDBShellCommandResolver.test.ts from jest to vitest.

mkrueger committed:
Address Cosmos DB Shell review feedback

mkrueger commented on src/extension.ts:
```diff
@@ -87,6 +92,9 @@ export async function activateInternal(
         ext.fileSystem = new DatabasesFileSystem();
 
+        const cosmosDBShellSupport: CosmosDBShellExtension = new CosmosDBShellExtension();
+        await cosmosDBShellSupport.activate();
```
Valid concern. Fixed by adding the CosmosDBShellExtension instance to context.subscriptions before activation so VS Code disposes its terminal listeners and tracked terminal state on deactivation/reload. Pushed in d499eb85.

mkrueger commented on src/cosmosDBShell/CosmosDBShellExtension.ts:
```diff
@@ -0,0 +1,1130 @@
+}
+
+/**
+ * Watches for the terminal closing shortly after creation (early exit).
+ * If the process exits quickly, logs the exit code and reason to the output channel.
+ */
+function watchForEarlyExit(terminal: vscode.Terminal): void {
+    const startTime = Date.now();
+    const listener = vscode.window.onDidCloseTerminal((closedTerminal) => {
+        listener.dispose();
+        if (closedTerminal === terminal && Date.now() - startTime < 5000) {
+            const exitCode = closedTerminal.exitStatus?.code;
+            const exitReason = closedTerminal.exitStatus?.reason;
+            ext.outputChannel.error(
+                `Cosmos DB Shell exited early.${exitCode !== undefined ? ` Exit code: ${exitCode}.` : ''}${exitReason !== undefined ? ` Reason: ${exitReason}.` : ''}`,
+            );
+        }
+    });
```
Valid concern. watchForEarlyExit now ignores close events for other terminals, disposes only when the tracked shell closes, and also cleans up the listener after the early-exit window. Pushed in d499eb85.

mkrueger replied:
Valid concern. watchForEarlyExit now ignores close events for other terminals, disposes only when the tracked shell closes, and also cleans up the listener after the early-exit window. Pushed in d499eb85.

mkrueger commented on src/cosmosDBShell/cosmosDBShellCommandResolver.ts:
```diff
@@ -0,0 +1,122 @@
+    return stripWrappingQuotes(trimmedShellPath);
+}
+
+function stripWrappingQuotes(value: string): string {
+    return (value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))
+        ? value.slice(1, -1)
+        : value;
+}
+
+function resolveWindowsCommand(command: string, env: NodeJS.ProcessEnv): string | undefined {
+    if (isExplicitPath(command)) {
+        return isFile(command) ? command : undefined;
+    }
```
Valid concern. Explicit Windows paths without extensions now use the same PATHEXT candidate probing as PATH lookup, with a focused resolver test covering the configured path case. Pushed in d499eb85.

mkrueger replied:
Valid concern. Explicit Windows paths without extensions now use the same PATHEXT candidate probing as PATH lookup, with a focused resolver test covering the configured path case. Pushed in d499eb85.

mkrueger commented on package.json:
```diff
@@ -608,6 +625,12 @@
           "when": "view =~ /azure(ResourceGroups|Workspace|FocusView)/ && viewItem =~ /treeitem[.]database(?![a-z.\\/])/i && viewItem =~ /experience[.](core)/i",
           "group": "1@3"
         },
+        {
+          "//": "[Database] Launch Cosmos DB Shell",
+          "command": "cosmosDB.launchCosmosDBShell",
+          "when": "view =~ /azure(ResourceGroups|Workspace|FocusView)/ && (viewItem =~ /treeitem[.]database(?![a-z.\\/])/i || viewItem =~ /treeitem[.]container(?![a-z.\\/])/i) && viewItem =~ /experience[.]core/i",
+          "group": "1@4"
+        },
```
Valid concern. Added vscodeDatabases.cosmosDBShellSupportEnabled to the context-menu when clause so the menu item matches the advertised gating. Pushed in d499eb85.

mkrueger commented on src/cosmosDBShell/CosmosDBShellExtension.ts:
```diff
@@ -0,0 +1,1130 @@
+ * entry-point for mongoClusters-related code. Activated from ./src/extension.ts
+ *
+ * We'll try to have everything related to mongoClusters-support managed from here.
+ * In case of a failure with this plan, this comment section will be updated.
```
Valid concern. Updated the file header to describe the Cosmos DB Shell responsibilities instead of the old mongoClusters text. Pushed in d499eb85.

mkrueger commented on src/cosmosDBShell/CosmosDBShellExtension.ts:
```diff
@@ -0,0 +1,1130 @@
+    if (!rawConnectionString) {
+        void vscode.window.showErrorMessage(l10n.t('Failed to extract the connection string from the selected node.'));
+        return;
+    }
```
Valid concern. The launch/connect paths now use rawEndpoint naming and the user-facing error says account endpoint, with localization regenerated. Pushed in d499eb85.

mkrueger commented on src/constants.ts:
```diff
@@ -38,6 +38,15 @@ export function getThemeAgnosticIconPath(iconName: string): IconPath {
     return Uri.file(icon);
 }
 
+export function getThemeAgnosticIconURI(iconName: string): IconPath {
+    const iconPath = Uri.joinPath(ext.context.extensionUri, 'resources', 'icons', 'theme-agnostic', iconName);
+    assert.ok(fs.existsSync(iconPath.fsPath), `Icon not found: ${iconPath.fsPath}`);
+    return {
+        light: iconPath,
+        dark: iconPath,
+    };
+}
```
Valid concern. getThemeAgnosticIconURI now logs the missing icon and falls back to a ThemeIcon instead of asserting in production. Pushed in d499eb85.

mkrueger commented on src/cosmosDBShell/CosmosDBShellExtension.ts:
```diff
@@ -0,0 +1,1130 @@
+
+// Create a singleton instance to access the context updater
+//const cosmosDBShellExt = new CosmosDBShellExtension();
+
```
Valid concern. Removed the stale commented-out singleton code. Pushed in d499eb85.

mkrueger committed:
Update localization bundle

mkrueger committed:
Keep Cosmos DB Shell menu visible before install
